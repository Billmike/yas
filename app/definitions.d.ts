interface INavigationProps {
  navigation?: {
    getParam: (paramName: string, defaultValue?: any) => any;
    navigate: (routeName: string, params?: any) => void;
  };
  route?: {
    params: any;
  };
}
