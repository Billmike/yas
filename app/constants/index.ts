export const HEADER_MAX_HEIGHT = 250;
export const HEADER_MIN_HEIGHT = 120;
export const CHARACTER_PROFILE_MAX_HEIGHT = 80;
export const CHARACTER_PROFILE_MIN_HEIGHT = 40;
