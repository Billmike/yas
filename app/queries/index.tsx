import { gql } from '@apollo/client';

export const GET_CHARACTERS = gql`
  query GetCharacters($page: Int, $name: String!) {
    characters(page: $page, filter: { name: $name }) {
      info {
        next
      }
      results {
        name
        image
        id
      }
    }
  }
`;

export const GET_CHARACTER = gql`
  query Character($id: ID!) {
    character(id: $id) {
      name
      species
      image
      gender
      episode {
        episode
        name
        air_date
      }
    }
  }
`;

export const SEARCH_CHARACTERS = gql`
  query GetCharacters($name: String!) {
    characters(filter: {name: $name}) {
      results {
        name
        image
        id
      }
    }
  }
`;
