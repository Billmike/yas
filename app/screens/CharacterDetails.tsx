/* eslint-disable react-native/no-inline-styles */
import * as React from 'react';
import {
  View,
  FlatList,
  Text,
  Image,
  StyleSheet,
  ActivityIndicator,
  ScrollView,
  Animated,
} from 'react-native';
import { useQuery } from '@apollo/client';
import { GET_CHARACTER } from '../queries';
import Error from '../components/ErrorView';
import {
  HEADER_MAX_HEIGHT,
  CHARACTER_PROFILE_MAX_HEIGHT,
  HEADER_MIN_HEIGHT,
  CHARACTER_PROFILE_MIN_HEIGHT,
} from '../constants';

interface ICharacterDetails extends INavigationProps { }

type EpisodeItem = {
  episode: string;
  name: string;
  air_date: string;
};

const styles = StyleSheet.create({
  header: {
    fontSize: 15,
    fontWeight: 'bold',
  },
  text: {
    fontSize: 14,
    marginTop: 2,
  },
  row: {
    flexDirection: 'row',
    marginBottom: 8,
    alignItems: 'center',
  },
  image: {
    height: 250,
    width: 250,
    alignSelf: 'center',
    marginBottom: 8,
    marginTop: 8,
  },
  container: {
    flex: 1,
    backgroundColor: '#fff',
    paddingHorizontal: 20,
  },
  imageHeader: {
    position: 'absolute',
    left: 0,
    right: 0,
  },
  imageView: {
    borderRadius: CHARACTER_PROFILE_MAX_HEIGHT / 2,
    overflow: 'hidden',
    borderWidth: 3,
    borderColor: '#fff',
    marginTop: HEADER_MAX_HEIGHT - CHARACTER_PROFILE_MAX_HEIGHT / 2,
  },
  overlay: {
    position: 'absolute',
    left: 0,
    right: 0,
    backgroundColor: 'rgba(1, 34, 36, 1)',
    width: '100%',
    height: '100%',
    opacity: 0.4,
  },
});

const Episodes = ({ episode, name, air_date }: EpisodeItem) => {
  return (
    <View style={{ marginTop: 15 }}>
      <Text style={{ marginBottom: 4 }}>{episode}</Text>
      <View style={{ flexDirection: 'row', marginBottom: 8 }}>
        <Text style={styles.header}>Title: </Text>
        <Text style={styles.text}>{name}</Text>
      </View>
      <View style={{ flexDirection: 'row' }}>
        <Text style={styles.header}>Air Date: </Text>
        <Text style={styles.text}>{air_date}</Text>
      </View>
    </View>
  );
};

const CharacterDetails = ({ route }: ICharacterDetails) => {
  const { loading, data, error } = useQuery(GET_CHARACTER, {
    variables: { id: route?.params.id },
    fetchPolicy: 'cache-first',
  });
  const [scrollY, setScrollY] = React.useState(new Animated.Value(0));
  const inputRange = [0, HEADER_MAX_HEIGHT - HEADER_MIN_HEIGHT];

  const headerHeight = scrollY.interpolate({
    inputRange,
    outputRange: [HEADER_MAX_HEIGHT, HEADER_MIN_HEIGHT],
    extrapolate: 'clamp',
  });

  const headerZIndex = scrollY.interpolate({
    inputRange,
    outputRange: [0, 1],
    extrapolate: 'clamp',
  });

  const profileImageHeight = scrollY.interpolate({
    inputRange,
    outputRange: [CHARACTER_PROFILE_MAX_HEIGHT, CHARACTER_PROFILE_MIN_HEIGHT],
    extrapolate: 'clamp',
  });

  const profileMarginTop = scrollY.interpolate({
    inputRange,
    outputRange: [
      HEADER_MAX_HEIGHT - CHARACTER_PROFILE_MAX_HEIGHT / 2,
      HEADER_MAX_HEIGHT,
    ],
    extrapolate: 'clamp',
  });

  const headerTitleBottom = scrollY.interpolate({
    inputRange: [
      0,
      HEADER_MAX_HEIGHT - HEADER_MIN_HEIGHT,
      HEADER_MAX_HEIGHT - HEADER_MIN_HEIGHT + 5 + CHARACTER_PROFILE_MIN_HEIGHT,
      HEADER_MAX_HEIGHT -
      HEADER_MIN_HEIGHT +
      5 +
      CHARACTER_PROFILE_MIN_HEIGHT +
      25,
    ],
    outputRange: [-20, -20, -20, 10],
    extrapolate: 'clamp',
  });

  return (
    <View style={styles.container}>
      <Animated.View
        style={[
          styles.imageHeader,
          { height: headerHeight, zIndex: headerZIndex },
        ]}>
        <Image
          source={{ uri: data?.character.image || '' }}
          style={{ flex: 1, height: undefined, width: undefined }}
        />
        <View style={styles.overlay} />
        <Animated.View
          style={{
            position: 'absolute',
            bottom: headerTitleBottom,
            left: 0,
            right: 0,
          }}>
          <Text
            style={{
              fontSize: 15,
              fontWeight: 'bold',
              textAlign: 'center',
              color: '#fff',
            }}>
            {data?.character.name}
          </Text>
        </Animated.View>
      </Animated.View>
      {error && <Error text="Error fetching your request. Please try again" />}
      {loading && <ActivityIndicator size="large" style={{ marginTop: 50 }} />}
      <ScrollView
        style={{ flex: 1 }}
        onScroll={Animated.event([
          {
            nativeEvent: {
              contentOffset: {
                y: scrollY,
              },
            },
          },
        ])}
        scrollEventThrottle={16}>
        <Animated.View
          style={[
            styles.imageView,
            {
              height: profileImageHeight,
              width: profileImageHeight,
              marginTop: profileMarginTop,
            },
          ]}>
          <Image
            source={{ uri: data?.character.image }}
            style={{ flex: 1, height: undefined, width: undefined }}
          />
        </Animated.View>
        <Text
          style={[
            styles.text,
            { fontSize: 25, fontWeight: 'bold', marginBottom: 10 },
          ]}>
          {data?.character.name}
        </Text>
        <View style={styles.row}>
          <Text style={styles.header}>Gender: </Text>
          <Text style={styles.text}>{data?.character.gender}</Text>
        </View>
        <View style={styles.row}>
          <Text style={styles.header}>Specie: </Text>
          <Text style={styles.text}>{data?.character.species}</Text>
        </View>
        <Text style={[styles.header]}>Episodes:</Text>
        <FlatList
          data={data?.character.episode}
          renderItem={({ item }) => <Episodes {...item} />}
          keyExtractor={(item, index) => `${index}`}
          contentContainerStyle={{ paddingBottom: 100 }}
        />
        <View style={{ height: 1000 }} />
      </ScrollView>
    </View>
  );
};

export default CharacterDetails;
