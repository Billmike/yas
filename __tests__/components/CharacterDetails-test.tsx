import React from 'react';
import { render } from '@testing-library/react-native';
import CharacterDetails from '../../app/screens/CharacterDetails';
import { MockedProvider } from '@apollo/client/testing';
import { GET_CHARACTER } from '../../app/queries';

const mocks = [
  {
    request: {
      query: GET_CHARACTER,
      variables: {
        id: '1'
      }
    },
    result: {
      data: {
        character: {
          name: 'Rick',
          gender: 'Male',
          species: 'Human',
          image: 'dummyImage',
          episode: [
            { episode: 'S01E01', name: 'Pilot', air_date: 'Nov 20, 2018' }
          ]
        }
      }
    }
  }
];

test('should render without error', () => {
  const { toJSON } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <CharacterDetails route={{ params: { id: '1' } }} />
    </MockedProvider>
  );
  expect(toJSON()).toMatchSnapshot()
})