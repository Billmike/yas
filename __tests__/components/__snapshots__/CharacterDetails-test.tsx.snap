// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without error 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingHorizontal": 20,
    }
  }
>
  <View
    style={
      Object {
        "height": 250,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "zIndex": 0,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "",
        }
      }
      style={
        Object {
          "flex": 1,
          "height": undefined,
          "width": undefined,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "rgba(1, 34, 36, 1)",
          "height": "100%",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "bottom": -20,
          "left": 0,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      />
    </View>
  </View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "marginTop": 50,
      }
    }
  />
  <RCTScrollView
    onScroll={[Function]}
    scrollEventThrottle={16}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "borderColor": "#fff",
            "borderRadius": 40,
            "borderWidth": 3,
            "height": 80,
            "marginTop": 210,
            "overflow": "hidden",
            "width": 80,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "flex": 1,
              "height": undefined,
              "width": undefined,
            }
          }
        />
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginTop": 2,
            },
            Object {
              "fontSize": 25,
              "fontWeight": "bold",
              "marginBottom": 10,
            },
          ]
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
            }
          }
        >
          Gender: 
        </Text>
        <Text
          style={
            Object {
              "fontSize": 14,
              "marginTop": 2,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
            }
          }
        >
          Specie: 
        </Text>
        <Text
          style={
            Object {
              "fontSize": 14,
              "marginTop": 2,
            }
          }
        />
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
            },
          ]
        }
      >
        Episodes:
      </Text>
      <RCTScrollView
        contentContainerStyle={
          Object {
            "paddingBottom": 100,
          }
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
      <View
        style={
          Object {
            "height": 1000,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
